File: Data Structures & Algorithms/balanced-binary-tree/submission-9.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isBalanced(TreeNode root) {
        int [] result = dfs(root);
        return result[0] == 1 ? true : false;
    }

    public int [] dfs(TreeNode root){
        if(root == null){
            return new int []{1, 0};
        }

        int [] resultLeft = dfs(root.left);
        int [] resultRight = dfs(root.right);

        boolean isBalanced = false;

        if(resultLeft[0] == 1 && resultRight[0] == 1 && Math.abs(resultLeft[1] - resultRight[1]) <= 1){
            isBalanced = true;
        }

        int maxDepth = Math.max(resultRight[1], resultLeft[1]) + 1;

        return new int [] {isBalanced ? 1 : 0, maxDepth};
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-9.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count_characters = {}
        most_frequent = 0
        result = 0

        left = 0

        for right in range(len(s)):
            count_characters[s[right]] = count_characters.get(s[right], 0) + 1

            most_frequent = max(count_characters[s[right]], most_frequent)

            while (right - left + 1) - most_frequent > k:
                count_characters[s[left]] = count_characters.get(s[left], 0) - 1
                left += 1
            
            result = max((right - left + 1), result)
        
        return result

